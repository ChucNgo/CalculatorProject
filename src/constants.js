export const colors = {
  black: "#000000",
  gray: "#DDDDDD"
};
export const keypadLeft = [
  ["7", "8", "9"],
  ["4", "5", "6"],
  ["1", "2", "3"],
  ["C", "0", "."]
];
export const keypadRight = ["/", "*", "+", "-", "="];
export const types = {
  CALCULATE: "CALCULATE",
  CLEAR: "CLEAR",
  SET_RESULT: "SET_RESULT"
};

export const evaluateExpression = (expression, total, data) => {
  if (/[\d]*[-+*/.]$/.exec(expression) && /[-+*/.]/.exec(data)) {
    expression = expression.slice(0, expression.length - 1);
  }

  if (["+", "/", "*", "%"].includes(data) && !expression) {
    return `${total}${data}`;
  }

  return `${!expression && total ? total : ""}${expression + data}`;
};

export const calculateTotal = expression => {
  const matched = new RegExp(
    "([\\d]+\\.?[\\d]*)?([-+/*][\\d]+\\.?[\\d]*)*"
  ).exec(expression);
  if (!matched) {
    return 0;
  }
  if (/^[*+\/]/.test(expression)) {
    return () => {
      throw new Error("Cannot start the expression with invalid operators");
    };
  }
  return eval(matched[0]);
};
