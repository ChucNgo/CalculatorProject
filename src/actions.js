import { types } from "./constants";

export const handleInputData = value => dispatch => {
  switch (value) {
    case "C":
      dispatch(clear());
      break;
    case "=":
      dispatch(setResult());
      break;
    default:
      dispatch(calculate(value));
      break;
  }
};

export const calculate = data => ({
  type: types.CALCULATE,
  data
});

export const clear = () => ({
  type: types.CLEAR
});

export const setResult = () => ({
  type: types.SET_RESULT
});
