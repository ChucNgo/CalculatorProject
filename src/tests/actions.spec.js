import { types } from "../constants";
import { calculate, clear, setResult, handleInputData } from "../actions";
import mockStore from "redux-mock-store";

describe("handle triggered actions", () => {
  it("creates a HANDLE INPUT DATA action", () => {
    const store = mockStore({ expression: "1+2-3*4", total: 0 });
    store.dispatch(handleInputData("="));
    expect(store.getActions()).toMatchSnapshot();

    store.dispatch(handleInputData("C"));
    expect(store.getActions()).toMatchSnapshot();

    store.dispatch(handleInputData("5"));
    expect(store.getActions()).toMatchSnapshot();
  });

  it("creates a CLEAR action", () => {
    expect(clear()).toEqual({
      type: types.CLEAR
    });
    expect(clear()).toMatchSnapshot();
  });

  it("creates a CALCULATE action", () => {
    expect(calculate(5)).toEqual({
      type: types.CALCULATE,
      data: 5
    });
    expect(calculate(5)).toMatchSnapshot();
  });

  it("creates a SET_RESULT action", () => {
    expect(setResult()).toEqual({
      type: types.SET_RESULT
    });
    expect(setResult()).toMatchSnapshot();
  });
});
