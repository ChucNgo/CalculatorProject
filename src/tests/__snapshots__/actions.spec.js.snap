// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handle triggered actions creates a CALCULATE action 1`] = `
Object {
  "data": 5,
  "type": "CALCULATE",
}
`;

exports[`handle triggered actions creates a CLEAR action 1`] = `
Object {
  "type": "CLEAR",
}
`;

exports[`handle triggered actions creates a HANDLE INPUT DATA action 1`] = `
Array [
  Object {
    "type": "SET_RESULT",
  },
]
`;

exports[`handle triggered actions creates a HANDLE INPUT DATA action 2`] = `
Array [
  Object {
    "type": "SET_RESULT",
  },
  Object {
    "type": "CLEAR",
  },
]
`;

exports[`handle triggered actions creates a HANDLE INPUT DATA action 3`] = `
Array [
  Object {
    "type": "SET_RESULT",
  },
  Object {
    "type": "CLEAR",
  },
  Object {
    "data": "5",
    "type": "CALCULATE",
  },
]
`;

exports[`handle triggered actions creates a SET_RESULT action 1`] = `
Object {
  "type": "SET_RESULT",
}
`;
